### PR TITLE
feat: Ceate description text view with "read more" option

### PR DESCRIPTION
**Pull request by ahmdyasser**

### CODE REVIEW

**Review by claramounir — APPROVED** [review]

File: UIDokan/Sources/UIDokan/ReusableViews/ReadMoreTextView.swift [hunk 1]
```diff
@@ -0,0 +1,438 @@
...
+
+extension String {
+    var length: Int {
+        return utf16.count
+    }
+}
+
+
+
+extension NSLayoutManager {
```
Comment (ahmdmhasn) [inline UIDokan/Sources/UIDokan/ReusableViews/ReadMoreTextView.swift:377]:
```suggestion
// MARK: - NSLayoutManager Private Helpers

private extension NSLayoutManager {
```

File: UIDokan/Sources/UIDokan/ReusableViews/ReadMoreTextView.swift [hunk 2]
```diff
@@ -0,0 +1,438 @@
...
+     */
+    public func boundingRectForCharacterRange(range aRange: NSRange, inTextContainer container: NSTextContainer) -> CGRect {
+        let glyphRange = self.glyphRange(forCharacterRange: aRange, actualCharacterRange: nil)
+        let boundingRect = self.boundingRect(forGlyphRange: glyphRange, in: container)
+        return boundingRect
+    }
+    
+}
+
+private extension UITextView {
```
Comment (ahmdmhasn) [inline UIDokan/Sources/UIDokan/ReusableViews/ReadMoreTextView.swift:401]:
```suggestion

// MARK: - UITextView Private Helpers

private extension UITextView {
```

File: UIDokan/Sources/UIDokan/ReusableViews/ReadMoreTextView.swift [hunk 3]
```diff
@@ -0,0 +1,438 @@
...
+    
+}
+
+extension String {
+    var length: Int {
+        return utf16.count
+    }
+}
+
+
```
Comment (ahmdmhasn) [inline UIDokan/Sources/UIDokan/ReusableViews/ReadMoreTextView.swift]:
```suggestion
// MARK: String Private Helpers

private extension String {
    var dataLength: Int {
        return utf16.count
    }
}
```

Comment (ahmdmhasn) [reply UIDokan/Sources/UIDokan/ReusableViews/ReadMoreTextView.swift]: TO remove confusion with count.

File: UIDokan/Sources/UIDokan/ReusableViews/ReadMoreTextView.swift [hunk 4]
```diff
@@ -0,0 +1,438 @@
+//
```
Comment (ahmdmhasn) [inline UIDokan/Sources/UIDokan/ReusableViews/ReadMoreTextView.swift]: It's preferred to add the source link.

File: Dokan/Dokan/Scenes/ProductDetails/ProductDetailsViewController.swift [hunk 5]
```diff
@@ -6,11 +6,13 @@
 //
 
 import UIKit
+import UIDokan
 
 class ProductDetailsViewController: UIViewController {
     
     // MARK: Outlets
-
+    @IBOutlet weak var descriptionTextView: ReadMoreTextView!
```
Comment (ahmdmhasn) [inline Dokan/Dokan/Scenes/ProductDetails/ProductDetailsViewController.swift]:
```suggestion
    @IBOutlet private weak var descriptionTextView: ReadMoreTextView!
```

**Comment by ahmdyasser** [comment]
@ahmdmhasn  🚀 Attention: You are about to see a piece of code art

File: UIDokan/Sources/UIDokan/ReusableViews/ReadMoreTextView.swift [hunk 6]
```diff
@@ -0,0 +1,442 @@
...
+            return false
+        } else if readLessText != nil && pointIsInTextRange(point: aPoint, range: readLessTextRange(), padding: readLessTextPadding) {
+            return true
+        }
+        return nil
+    }
+    
+}
+
+extension String {
```
Comment (ahmdmhasn) [inline UIDokan/Sources/UIDokan/ReusableViews/ReadMoreTextView.swift]:
```suggestion
private extension String {
```

File: UIDokan/Sources/UIDokan/ReusableViews/ReadMoreTextView.swift [hunk 7]
```diff
@@ -0,0 +1,442 @@
...
+        if needsTrim() && pointIsInTextRange(point: aPoint, range: readMoreTextRange(), padding: readMoreTextPadding) {
+            return false
+        } else if readLessText != nil && pointIsInTextRange(point: aPoint, range: readLessTextRange(), padding: readLessTextPadding) {
+            return true
+        }
+        return nil
+    }
+    
+}
+
```
Comment (ahmdmhasn) [inline UIDokan/Sources/UIDokan/ReusableViews/ReadMoreTextView.swift:365]:
```suggestion

// MARK: String Private Helpers

```

**Review by ahmdmhasn — APPROVED** [review]

**Pull request merged**